### ISSUE TITLE
Physics-informed loss with div-B=0 regularizer

### ISSUE THREAD

**Issue by chadlindell** [chadlindell/terrascry#541]
## Summary

Implement custom loss functions including a physics-informed ∇·B=0 regularizer for the U-Net training pipeline.

## Background

AI counsel consensus: use ∇·B=0 as a **loss regularizer only**, not full PDE-constrained PINNs. Full Maxwell's equations as loss has limited payoff for non-unique magnetic inversion in real-time. The divergence-free constraint is physically universal and computationally cheap.

## Loss Components

```
L_total = L_denoise + α · L_classify + λ · ||div(B)||²
```

1. **L_denoise**: MSE between predicted noise residual and true noise (clean − noisy)
2. **L_classify**: Weighted cross-entropy for 5-class classification
3. **∇·B=0 penalty**: Finite-difference divergence on predicted denoised field
   - `div(B) = ∂Bx/∂x + ∂By/∂y` (2D gradient components)
   - Computed via 2D finite differences on output grid
   - Penalizes predictions that violate Maxwell's equations

## Requirements

- Configurable weights: α (classification), λ (physics regularizer)
- Finite difference divergence computation (differentiable, GPU-compatible)
- Class weights for cross-entropy (inversely proportional to frequency)
- Logging: individual loss components tracked separately
- Unit tests: verify gradient flow through all components

## Dependencies

- #539 (U-Net architecture)
- #540 (classification head)

---

`needs-planning` — requires AI counsel consensus before implementation.

**Epic:** Real-Time Interpretation Engine
**Phase:** 2 — Model Architecture & Training
